src/client/particles.cpp

```diff
@@ -31,635 +29,808 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+				if (collision_removal) {
+					p.endTime = 0;
+				} else {
+					p.pos = position - camera_offset;
+					p.vector = velocity * 1e-3f;
+				}
+			}
 		}
-		m_box.addInternalPoint(vertex.Pos);
-		vertex.Pos += m_pos*BS - intToFloat(camera_offset, BS);
 	}
-}
-
-/*
-	ParticleSpawner
-*/
 
-ParticleSpawner::ParticleSpawner(
-	IGameDef *gamedef,
-	LocalPlayer *player,
-	u16 amount,
-	float time,
-	v3f minpos, v3f maxpos,
-	v3f minvel, v3f maxvel,
-	v3f minacc, v3f maxacc,
-	float minexptime, float maxexptime,
-	float minsize, float maxsize,
-	bool collisiondetection,
-	bool collision_removal,
-	bool object_collision,
-	u16 attached_id,
-	bool vertical,
-	video::ITexture *texture,
-	const struct TileAnimationParams &anim,
-	u8 glow,
-	ParticleManager *p_manager
-):
-	m_particlemanager(p_manager)
+	virtual irr::scene::E_PARTICLE_AFFECTOR_TYPE getType() const
+	{
+		return irr::scene::EPAT_NONE;
+	}
+private:
+	ClientEnvironment *env;
+	u32 last_time;
+	const bool collision_removal;
+	const bool object_collision;
+	const f32 bounce_fraction;
+	const f32 bounce_threshold;
+};
+
+class AccelerationAffector : public irr::scene::IParticleAffector
 {
-	m_gamedef = gamedef;
-	m_player = player;
-	m_amount = amount;
-	m_spawntime = time;
-	m_minpos = minpos;
-	m_maxpos = maxpos;
-	m_minvel = minvel;
-	m_maxvel = maxvel;
-	m_minacc = minacc;
-	m_maxacc = maxacc;
-	m_minexptime = minexptime;
-	m_maxexptime = maxexptime;
-	m_minsize = minsize;
-	m_maxsize = maxsize;
-	m_collisiondetection = collisiondetection;
-	m_collision_removal = collision_removal;
-	m_object_collision = object_collision;
-	m_attached_id = attached_id;
-	m_vertical = vertical;
-	m_texture = texture;
-	m_time = 0;
-	m_animation = anim;
-	m_glow = glow;
-
-	for (u16 i = 0; i<=m_amount; i++)
+public:
+	AccelerationAffector(const v3f &minacc, const v3f &maxacc)
+		: last_time(0)
+	{
+		disabled = minacc.getLengthSQ() <= 1e-6 && maxacc.getLengthSQ() <= 1e-6;
+		acc = random_v3f(minacc,maxacc);
+	}
+	AccelerationAffector(const v3f &acc)
+		: acc(acc), disabled(acc.getLengthSQ() == 0.f), last_time(0)
 	{
-		float spawntime = (float)rand()/(float)RAND_MAX*m_spawntime;
-		m_spawntimes.push_back(spawntime);
 	}
-}
 
-void ParticleSpawner::spawnParticle(ClientEnvironment *env, float radius,
-	bool is_attached, const v3f &attached_pos, float attached_yaw)
+	virtual void affect(u32 now, irr::scene::SParticle *particlearray, u32 count)
+	{
+		if (last_time == 0) {
+			last_time = now;
+			return;
+		}
+		float dtime = (now - last_time) * 1e-3f;
+		last_time = now;
+		for (u32 i = 0; i < count; ++i) {
+			particlearray[i].vector += dtime * acc * BS * 1e-3f;
+		}
+	}
+
+	virtual irr::scene::E_PARTICLE_AFFECTOR_TYPE getType() const
+	{
+		return irr::scene::EPAT_NONE;
+	}
+private:
+	v3f acc;
+	bool disabled;
+	u32 last_time;
+};
+class CustomEmitter : public irr::scene::IParticleEmitter
+{
+public:
+	// unused pure virtual methods
+	virtual void setDirection(const core::vector3df& /* newDirection */) {}
```

Comment:
Haven't checked, but if these are from IParticleEmitter then please use override instead like this instead:

```c++
void setDirection(const core::vector3df& /* newDirection */) override {}
```